comp/core/secrets/secretsimpl/rotating_ndrecords.go

```diff
@@ -0,0 +1,299 @@
+		if err = json.Unmarshal(line, &rec); err == nil {
+			if !rec.Time.IsZero() && now.Sub(rec.Time) <= r.cfg.retention {
+				r.firstEntry = rec.Time
+				break
+			}
+		}
+	}
+	tmpForRename, _ := os.CreateTemp("", "replace")
+	if !r.firstEntry.IsZero() {
+		// if we found an entry to keep, write it first
+		json.NewEncoder(tmpForRename).Encode(rec) //nolint:errcheck
+	}
+	// write the remainder of the file
+	if _, err := io.Copy(tmpForRename, rd); err != nil {
+		return err
+	}
+	tmpForRename.Close()
+	f.Close()
+	return os.Rename(tmpForRename.Name(), r.filename)
+}
+
+// remove any old rotated files that are past the retention time
+func (r *rotatingNDRecords) removeOldFiles(t time.Time) error {
+	for _, filename := range r.RotatedFiles() {
+		if stat, err := os.Stat(filename); err == nil {
+			if t.Sub(stat.ModTime()) > r.cfg.retention {
+				log.Infof("removing old rotated file '%s'", filename)
+				os.Remove(filename)
+				r.oldestFileMtime = nil
+			} else {
+				break
+			}
+		}
+	}
+	return nil
+}
+
+// rotate the current file to the next available name
+func (r *rotatingNDRecords) rotateFile() {
+	rotateDestFilename, err := nextRotateFilename(r.filename, r.cfg.spacer)
+	if err != nil {
+		log.Errorf("could not find rotation filename: %s", err)
+		return
+	}
+	if _, err := os.Stat(rotateDestFilename); errors.Is(err, os.ErrNotExist) {
+		if err := os.Rename(r.filename, rotateDestFilename); err != nil {
+			log.Errorf("could not rotate file: %s", err)
+			return
+		}
+		log.Infof("renamed large file '%s' to '%s'", r.filename, rotateDestFilename)
+		r.oldestFileMtime = nil
+	}
+}
+
+// nextRotateFilename calculates filename that should be used for next rotation
+func nextRotateFilename(filename string, spacer int) (string, error) {
+	dir := filepath.Dir(filename)
+	re, err := buildRotationRegex(filename, spacer)
+	if err != nil {
+		log.Error(err)
+	}
+
+	maxSpacerNum := -1
+	entries, err := os.ReadDir(dir)
+	if err != nil {
+		return "", err
+	}
+	for _, ent := range entries {
+		match := re.FindStringSubmatch(ent.Name())
+		if len(match) > 0 {
+			spacerNum, err := strconv.Atoi(match[1])
+			if err == nil && spacerNum > maxSpacerNum {
+				maxSpacerNum = spacerNum
+			}
+		}
+	}
+
+	fmtPattern, err := buildRotationFmtPattern(filename, spacer)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf(fmtPattern, maxSpacerNum+1), nil
+}
+
+func buildRotationRegex(filename string, spacer int) (*regexp.Regexp, error) {
+	base := filepath.Base(filename)
+	ext := filepath.Ext(base)
+	basenoext := strings.TrimSuffix(base, ext)
+	// add escape "\" for the "." at the start of the extension, if needed
+	maybeEscape := "\\"
+	if ext == "" {
+		maybeEscape = ""
+	}
+	// build a regex that matches rotating files
+	// for example, a filename like "records.ndjson" with spacer=6
+	// would build the regex "records\.(\d{6})\.ndjson"
+	pattern := fmt.Sprintf("%s\\.(\\d{%d})%s%s", basenoext, spacer, maybeEscape, ext)
+	return regexp.Compile(pattern)
+}
```

Comment:
This will generate an invalid regex as soon as the `filename` contains any character that has a special meaning for regex, `.`, `*`, `(`, `[`, etc. All those special characters need to be properly escaped.

```suggestion
func buildRotationRegex(filename string, spacer int) (*regexp.Regexp, error) {
	base := filepath.Base(filename)
	ext := filepath.Ext(base)
	basenoext := strings.TrimSuffix(base, ext)
	// build a regex that matches rotating files
	// for example, a filename like "records.ndjson" with spacer=6
	// would build the regex "records\.(\d{6})\.ndjson"
	pattern := fmt.Sprintf(`%s\.(\d{%d})%s", regexp.QuoteMeta(basenoext), spacer, regexp.QuoteMeta(ext))
	return regexp.Compile(pattern)
}
```

Reply:
This is much better, thanks!